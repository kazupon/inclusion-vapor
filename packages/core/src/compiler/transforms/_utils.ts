import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
