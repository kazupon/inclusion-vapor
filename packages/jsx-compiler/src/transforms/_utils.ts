import { parse } from '@babel/parser'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, JSXElement, JSXFragment } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const {
      body: [statement]
    } = parse(source, {
      sourceType: 'module',
      plugins: ['jsx']
    }).program
    let children!: JSXElement[] | JSXFragment['children']
    if (statement.type === 'ExpressionStatement') {
      children =
        statement.expression.type === 'JSXFragment'
          ? statement.expression.children
          : (statement.expression.type === 'JSXElement'
            ? [statement.expression]
            : [])
    }

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children,
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
