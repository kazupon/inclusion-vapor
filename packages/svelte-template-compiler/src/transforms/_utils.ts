import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
