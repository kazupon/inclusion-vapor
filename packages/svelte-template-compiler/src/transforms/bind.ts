// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vBind.ts

import {
  ErrorCodes,
  NodeTypes,
  createCompilerError,
  createSimpleExpression
} from '@vue-vapor/compiler-dom'
import { camelize, extend } from '@vue-vapor/shared'
import { isReservedProp, resolveExpression } from './utils.ts'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type { TransformContext } from './context.ts'
import type { DirectiveTransform } from './types.ts'

// same-name shorthand - :arg is expanded to :arg="arg"
export function normalizeBindShorthand(
  arg: SimpleExpressionNode,
  context: TransformContext
): SimpleExpressionNode {
  if (arg.type !== NodeTypes.SIMPLE_EXPRESSION || !arg.isStatic) {
    // only simple expression is allowed for same-name shorthand
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_BIND_INVALID_SAME_NAME_ARGUMENT, arg.loc)
    )
    return createSimpleExpression('', true, arg.loc)
  }

  const propName = camelize(arg.content)
  const exp = createSimpleExpression(propName, false, arg.loc)
  exp.ast = null // eslint-disable-line unicorn/no-null
  return exp
}

export const transformVBind: DirectiveTransform = (dir, _node, context) => {
  const { loc, modifiers } = dir
  let { exp } = dir
  let arg = dir.arg!
  const modifiersString = new Set(modifiers.map(s => s.content))

  if (!exp) {
    exp = normalizeBindShorthand(arg, context)
  }
  if (!exp.content.trim()) {
    context.options.onError(createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, loc))
    exp = createSimpleExpression('', true, loc)
  }

  exp = resolveExpression(exp)
  arg = resolveExpression(arg)

  if (arg.isStatic && isReservedProp(arg.content)) {
    return
  }

  let camel = false
  if (modifiersString.has('camel')) {
    if (arg.isStatic) {
      arg = extend({}, arg, { content: camelize(arg.content) })
    } else {
      camel = true
    }
  }

  return {
    key: arg,
    value: exp,
    loc,
    runtimeCamelize: camel,
    modifier: modifiersString.has('prop') ? '.' : modifiersString.has('attr') ? '^' : undefined
  }
}
