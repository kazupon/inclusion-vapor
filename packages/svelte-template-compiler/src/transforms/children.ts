// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { DynamicFlag, IRNodeTypes, isSvelteText } from '../ir/index.ts'
import { transformNode } from '../transform.ts'

import type { IRDynamicInfo, SvelteElement } from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'InlineComponent' ||
    node.type === 'Slot'
  if (node.type !== 'Element' && !isFragment) {
    return
  }
  processChildren(node, context, isFragment)
}

export function processChildren(
  node: Parameters<NodeTransform>[0],
  context: Parameters<NodeTransform>[1],
  isFragment: boolean
): ReturnType<NodeTransform> {
  const children = node.children || []

  // normalize svelte text
  for (const [index, child] of [...children].entries()) {
    if (isSvelteText(child) && !child.data.trim()) {
      child.data = ' '
      if (!index) {
        children.splice(0, 1)
      } else if (index === children.length) {
        children.splice(-1, 1)
      }
    }
  }

  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<SvelteElement>)
  }
}

function processDynamicChildren(context: TransformContext<SvelteElement>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
