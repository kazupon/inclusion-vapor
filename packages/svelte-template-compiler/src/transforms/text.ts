// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformText.ts

import { createSimpleExpression } from '@vue-vapor/compiler-dom'
import {
  convertToSourceLocation,
  DynamicFlag,
  IRNodeTypes,
  isSvelteElement,
  isSvelteMustacheTag,
  isSvelteText
} from '../ir/index.ts'
import {
  getLiteralExpressionValue,
  isConstantExpression,
  resolveSimpleExpression
} from './utils.ts'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type {
  RootNode,
  SvelteElement,
  SvelteMustacheTag,
  SvelteTemplateNode,
  SvelteText
} from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

type TextLike = SvelteText | SvelteMustacheTag

const seen = new WeakMap<TransformContext<RootNode>, WeakSet<SvelteTemplateNode | RootNode>>()

export const transformText: NodeTransform = (node, context) => {
  if (!seen.has(context.root)) {
    seen.set(context.root, new WeakSet())
  }

  if (seen.get(context.root)!.has(node)) {
    context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
    return
  }

  const children = node.children || []
  if (isSvelteElement(node) && isAllTextLike(children)) {
    // TODO: should more tested for component
    processTextLikeContainer(children, context as TransformContext<SvelteElement>)
  } else if (isSvelteMustacheTag(node)) {
    // TODO: should more tested for MustacheTag expression
    // e.g. ConditionalExpression, LogicalExpression and CallExpression ...
    processTextLike(context as TransformContext<SvelteMustacheTag>)
  } else if (isSvelteText(node)) {
    context.template += node.data
  }
}

function processTextLike(context: TransformContext<SvelteMustacheTag>): void {
  const children = context.parent!.node.children || []
  const nexts = children.slice(context.index)
  const idx = nexts.findIndex(n => !isTextLike(n))
  const nodes = (idx > -1 ? nexts.slice(0, idx) : nexts) as Array<TextLike>
  const id = context.reference()
  const values = nodes.map(node => createTextLikeExpression(node, context))

  context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  context.registerOperation({
    type: IRNodeTypes.CREATE_TEXT_NODE,
    id,
    values,
    // eslint-disable-next-line unicorn/no-array-callback-reference
    effect: !values.every(isConstantExpression) && !context.inVOnce
  })
}

function processTextLikeContainer(
  children: TextLike[],
  context: TransformContext<SvelteElement>
): void {
  const values = children.map(child => createTextLikeExpression(child, context))
  // eslint-disable-next-line unicorn/no-array-callback-reference
  const literals = values.map(getLiteralExpressionValue)
  if (literals.every(l => l != undefined)) {
    context.childrenTemplate = literals.map(String)
  } else {
    context.registerEffect(values, {
      type: IRNodeTypes.SET_TEXT,
      element: context.reference(),
      values
    })
  }
}

function createTextLikeExpression(node: TextLike, context: TransformContext): SimpleExpressionNode {
  seen.get(context.root)!.add(node)
  // prettier-ignore
  return isSvelteText(node)
    ? createSimpleExpression(node.data, true, convertToSourceLocation(node, (node as unknown as { raw: string }).raw || node.data))
    : resolveSimpleExpression(node, context as TransformContext<SvelteMustacheTag>)
}

function isAllTextLike(children: SvelteTemplateNode[]): children is TextLike[] {
  return (
    children.length > 0 &&
    // eslint-disable-next-line unicorn/no-array-callback-reference
    children.every(isTextLike) &&
    // at least one an mustache tag
    children.some(n => isSvelteMustacheTag(n))
  )
}

function isTextLike(node: SvelteTemplateNode): node is TextLike {
  // TODO: should more tested for MustacheTag expression
  // e.g. ConditionalExpression, LogicalExpression and CallExpression ...
  return isSvelteText(node) || isSvelteMustacheTag(node)
}
