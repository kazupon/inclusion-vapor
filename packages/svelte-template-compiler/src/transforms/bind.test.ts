import { NodeTypes } from '@vue-vapor/compiler-dom'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'
import { expect, test } from 'vitest'
import { DynamicFlag, IRNodeTypes } from '../ir/index.ts'
import { makeCompile } from './_utils.ts'
import { transformVBind } from './bind.ts'
import { transformChildren } from './children.ts'
import { transformElement } from './element.ts'
import { transformText } from './text.ts'

const compileWithVBind = makeCompile({
  nodeTransforms: [transformText, transformElement, transformChildren],
  directiveTransforms: {
    bind: transformVBind
  }
})

test('mustache basic', () => {
  const { ir, code, vaporHelpers } = compileWithVBind('<div id={id} />')
  const expectedResult = vaporCompile('<div :id="id" />', { prefixIdentifiers: true })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')
  expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)

  expect(code).contains('_setDynamicProp(n0, "id", _ctx.id, true)')

  expect(ir.block.dynamic.children[0]).toMatchObject({
    id: 0,
    flags: DynamicFlag.REFERENCED
  })
  expect(ir.template).toEqual(['<div></div>'])
  expect(ir.block.effect).lengthOf(1)
  expect(ir.block.effect[0].expressions).lengthOf(1)
  expect(ir.block.effect[0].operations).lengthOf(1)
  expect(ir.block.effect[0]).toMatchObject({
    expressions: [
      {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'id',
        isStatic: false
      }
    ],
    operations: [
      {
        type: IRNodeTypes.SET_PROP,
        element: 0,
        prop: {
          key: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'id',
            isStatic: true,
            loc: {
              // TODO:
              // start: { line: 1, column: 13, offset: 12 },
              // end: { line: 1, column: 15, offset: 14 },
              source: 'id'
            }
          },
          values: [
            {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: 'id',
              isStatic: false,
              loc: {
                source: 'id'
                // TODO:
                // start: { line: 1, column: 17, offset: 16 },
                // end: { line: 1, column: 19, offset: 18 },
              }
            }
          ],
          loc: {
            // TODO:
            // start: { column: 6, line: 1, offset: 5 },
            // end: { column: 20, line: 1, offset: 19 },
            // TODO: we want to map for svelte code correctly...
            // source: 'id={id}',
          },
          runtimeCamelize: false
        }
      }
    ]
  })
})

test('mustache shorthand', () => {
  const { ir, code, vaporHelpers } = compileWithVBind('<div {id} />')
  const expectedResult = vaporCompile('<div :id />', { prefixIdentifiers: true })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')
  expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)

  expect(code).contains('_setDynamicProp(n0, "id", _ctx.id, true)')
  expect(ir.block.effect[0].operations[0]).toMatchObject({
    type: IRNodeTypes.SET_PROP,
    prop: {
      key: {
        content: `id`,
        isStatic: true
      },
      values: [
        {
          content: `id`,
          isStatic: false
        }
      ]
    }
  })
})

test('attribute camel case', () => {
  const { code, vaporHelpers } = compileWithVBind('<div camel-case={value1} />')
  const expectedResult = vaporCompile('<div :camel-case="value1" /> ', {
    prefixIdentifiers: true
  })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')
  expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)

  expect(code).contains('_setDynamicProp(n0, "camel-case", _ctx.value1, true)')
})

test('class expression binding', () => {
  const { code, ir } = compileWithVBind(`<div class={ isActive ? 'active' : '' } />`)
  const expectedResult = vaporCompile(`<div :class="isActive ? 'active' : ''" />`, {
    prefixIdentifiers: true
  })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(code).contains(`_renderEffect(() => _setClass(n0, _ctx.isActive ? 'active' : '', true))`)
  expect(ir.block.effect[0]).toMatchObject({
    expressions: [
      {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: `isActive ? 'active' : ''`,
        isStatic: false
      }
    ],
    operations: [
      {
        type: IRNodeTypes.SET_PROP,
        element: 0,
        prop: {
          key: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'class',
            isStatic: true,
            loc: {
              // TODO:
              // source: 'class'
            }
          },
          values: [
            {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: `isActive ? 'active' : ''`,
              isStatic: false,
              loc: {
                source: `isActive ? 'active' : ''`
                // TODO:
              }
            }
          ],
          loc: {
            // TODO:
          },
          runtimeCamelize: false
        }
      }
    ]
  })
})

test('class dynamic binding', () => {
  const { code, ir } = compileWithVBind(`<div class:active={ isActive } />`)
  const expectedResult = vaporCompile(`<div :class="{ active: isActive }" />`, {
    prefixIdentifiers: true
  })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(code).contains(`_renderEffect(() => _setClass(n0, { active: _ctx.isActive }, true))`)
  expect(ir.block.effect[0]).toMatchObject({
    expressions: [
      {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: `{ active: isActive }`,
        isStatic: false
      }
    ],
    operations: [
      {
        type: IRNodeTypes.SET_PROP,
        element: 0,
        prop: {
          key: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'class',
            isStatic: true,
            loc: {
              // TODO:
              source: 'class'
            }
          },
          values: [
            {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: `{ active: isActive }`,
              isStatic: false,
              loc: {
                // TODO: also need to map for svelte code
                // source: `{ isActive }`,
              }
            }
          ],
          loc: {
            // TODO: also need to map for svelte code
            // source: `class:active={ isActive }`,
          },
          runtimeCamelize: false
        }
      }
    ]
  })
})

test('class shorthand binding', () => {
  const { code, ir } = compileWithVBind(`<div class:active />`)
  const expectedResult = vaporCompile(`<div :class="{ active }" />`, {
    prefixIdentifiers: true
  })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(code).contains(`_renderEffect(() => _setClass(n0, { active: _ctx.active }, true))`)
  expect(ir.block.effect[0]).toMatchObject({
    expressions: [
      {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: `{ active: active }`,
        isStatic: false
      }
    ],
    operations: [
      {
        type: IRNodeTypes.SET_PROP,
        element: 0,
        prop: {
          key: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'class',
            isStatic: true,
            loc: {
              // TODO:
              source: 'class'
            }
          },
          values: [
            {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: `{ active: active }`,
              isStatic: false,
              loc: {
                // TODO: also need to map for svelte code
                // source: `class:active`,
              }
            }
          ],
          loc: {
            // TODO: also need to map for svelte code
            // source: `class:active`,
          },
          runtimeCamelize: false
        }
      }
    ]
  })
})

test.todo('multiple class binding', () => {
  const { code, ir: _ } = compileWithVBind(
    `<div class="static" class:active class:inactive={!active} class:isAdmin />`
  )
  const expectedResult = vaporCompile(
    `<div class="static" :class="{ active, inactive: !active, isAdmin }" />`,
    {
      prefixIdentifiers: true
    }
  )

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(code).contains(
    `_renderEffect(() => _setClass(n2, ["static", { active, inactive: !active, isAdmin }]))`
  )
})

test('style expression binding', () => {
  const source1 = `<div style:color={myColor}>color</div>`
  const source2 = `<div :style="{ color: myColor }">color</div>`

  const { code, ir } = compileWithVBind(source1)
  const expectedResult = vaporCompile(source2, { prefixIdentifiers: true })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(ir.template).toEqual(['<div>color</div>'])

  expect(code).contains(`_renderEffect(() => _setStyle(n0, { color: _ctx.myColor }, true))`)
})

test('style shorthand binding', () => {
  const source1 = `<div style:color>color</div>`
  const source2 = `<div :style="{ color }">color</div>`

  const { code, ir } = compileWithVBind(source1)
  const expectedResult = vaporCompile(source2, { prefixIdentifiers: true })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(ir.template).toEqual(['<div>color</div>'])

  expect(code).contains(`_renderEffect(() => _setStyle(n0, { color: _ctx.color }, true))`)
})

test.todo('style `imporant` modifier', () => {
  const source1 = `<div style:color|important="red">modifier</div>`
  const source2 = `<div :style="{ color: 'red !important' }">modifier</div>`

  const { code, ir } = compileWithVBind(source1)
  const expectedResult = vaporCompile(source2, { prefixIdentifiers: true })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(code).contains(`_renderEffect(() => _setStyle(n0, { color: 'red !important' }))`)

  expect(ir.template).toEqual(['<div>modifier</div>'])
})

test.todo('multiple style binding', () => {
  const source1 = `<div style:color style:width="12rem" style:background-color={darkMode ? 'black' : 'white'}>multiple</div>`
  const source2 = `<div :style="{ color, width: '12rem', 'background-color': { darkMode ? 'black' : 'white' }">multiple</div>`

  const { code, ir } = compileWithVBind(source1)
  const expectedResult = vaporCompile(source2, { prefixIdentifiers: true })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(code).contains(
    `_renderEffect(() => _setStyle(n1, { color, width: '12rem', 'background-color': { darkMode ? 'black' : 'white' }))`
  )

  expect(ir.template).toEqual(['<div>multiple</div>'])
})

test.todo('style override', () => {
  const source1 = `<div style="color: blue;" style:color="red">This will be red</div>`
  const source2 = `<div style="color: blue;" :style="{ color: 'red' }">This will be red</div>`

  const { code, ir } = compileWithVBind(source1)
  const expectedResult = vaporCompile(source2, { prefixIdentifiers: true })

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(code).contains(`_renderEffect(() => _setStyle(n3, ["color: blue;", { color: 'red' }]))`)

  expect(ir.template).toEqual(['<div>This will be red</div>'])
})
