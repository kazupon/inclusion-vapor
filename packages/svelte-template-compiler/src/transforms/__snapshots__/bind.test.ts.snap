// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`attribute camel case > svelte 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["camel-case"])
  _renderEffect(() => _setDynamicProp(n0, "camel-case", _ctx.value1, true))
  return n0
}"
`;

exports[`attribute camel case > vue 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["camel-case"])
  _renderEffect(() => _setDynamicProp(n0, "camel-case", _ctx.value1, true))
  return n0
}"
`;

exports[`class dynamic binding > svelte 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, { active: _ctx.isActive }, true))
  return n0
}"
`;

exports[`class dynamic binding > vue 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, { active: _ctx.isActive }, true))
  return n0
}"
`;

exports[`class expression binding > svelte 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, _ctx.isActive ? 'active' : '', true))
  return n0
}"
`;

exports[`class expression binding > vue 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, _ctx.isActive ? 'active' : '', true))
  return n0
}"
`;

exports[`class shorthand binding > svelte 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, { active: _ctx.active }, true))
  return n0
}"
`;

exports[`class shorthand binding > vue 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, { active: _ctx.active }, true))
  return n0
}"
`;

exports[`multiple class binding > svelte 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, ["static", { active: _ctx.active }, { inactive: !_ctx.active }, { isAdmin: _ctx.isAdmin }], true))
  return n0
}"
`;

exports[`multiple class binding > vue 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, ["static", { active: _ctx.active, inactive: !_ctx.active, isAdmin: _ctx.isAdmin }], true))
  return n0
}"
`;

exports[`multiple class binding with scoped css > svelte 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setClass as _setClass, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["class"])
  _renderEffect(() => _setClass(n0, ["static _vapor_", { active: _ctx.active }, { inactive: !_ctx.active }, { isAdmin: _ctx.isAdmin }], true))
  return n0
}"
`;

exports[`mustache basic > svelte 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["id"])
  _renderEffect(() => _setDynamicProp(n0, "id", _ctx.id, true))
  return n0
}"
`;

exports[`mustache basic > vue 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["id"])
  _renderEffect(() => _setDynamicProp(n0, "id", _ctx.id, true))
  return n0
}"
`;

exports[`mustache shorthand > svelte 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["id"])
  _renderEffect(() => _setDynamicProp(n0, "id", _ctx.id, true))
  return n0
}"
`;

exports[`mustache shorthand > vue 1`] = `
"import { setInheritAttrs as _setInheritAttrs, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(["id"])
  _renderEffect(() => _setDynamicProp(n0, "id", _ctx.id, true))
  return n0
}"
`;
